### ISSUE TITLE
CA1062 false positive when a generic null checker method is used

### ISSUE THREAD

**Issue by ggreig** [dotnet/roslyn-analyzers#2504]
#### Analyzer package

[Microsoft.CodeAnalysis.FxCopAnalyzers](https://www.nuget.org/packages/Microsoft.CodeAnalysis.FxCopAnalyzers)

#### Package Version

v2.9.0, v2.9.1, v2.9.2

#### Diagnostic ID

[CA1062: Validate arguments of public methods](https://docs.microsoft.com/en-gb/visualstudio/code-quality/ca1062-validate-arguments-of-public-methods)

#### Repro steps

1. Enable CA1602
2. Analyze the following code (which is slightly modified from an example on the Diagnostic ID page):
```
public class Person
    {
        public string Name { get; private set; }
        public int Age { get; private set; }

        public Person(string name, int age)
        {
            Name = name;
            Age = age;
        }

        // Copy constructor
        public Person(Person other)
        {
            PassThroughNonNull(other);
            Name = other.Name;
            Age = other.Age;
        }

        /*
        // Null check method
        private static Person PassThroughNonNull(Person person)
        {
            if (person == null)
            {
                throw new ArgumentNullException("person");
            }

            return person;
        }
        */

        // Generic null check method
        private static T PassThroughNonNull<T>(T person)
        {
            if (person == null)
            {
                throw new ArgumentNullException("person");
            }

            return person;
        }
 }
```

#### Expected behaviour

No warning, since `other` is checked for `null` by the generic `PassThroughNull` method.

#### Actual behaviour

CA1062 is reported against `Name = other.Name;`

#### Further details

There are two significant changes above from the example code on the Diagnostic ID page. The pass through null check is being applied in the body of the constructor, rather than on parameters in a chained call to another constructor; and the pass through null check has been modified to accept a generic input rather than being a specific implementation for the Person class. We use such a generic null check widely throughout our codebase, and it satisfied pre-Roslyn versions of Code Analysis, so this failure is a serious one for us.

The null check from the Microsoft example (passes CA1062, but is too type-specific) has been left commented out in the example for convenience of comparison. The generic null check we've replaced it with seems to satisfy CA1062 when used in a chained call to another constructor but not when used in a method body such as the constructor in the example above.

**Comment by mavasani** [comment]
https://dotnet.myget.org/feed/roslyn-analyzers/package/nuget/Microsoft.CodeAnalysis.FxCopAnalyzers/2.9.4-beta1.19317.5+45537f7f should have the fix
